File: src/main/java/orderbook/orderbook/commands/OrderBookCollectCommand.java
package orderbook.orderbook.commands;

import orderbook.orderbook.Order;
import orderbook.orderbook.OrderBook;
import orderbook.orderbook.OrderManager;
import orderbook.orderbook.TradeInventoryManager;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.PlayerInventory;
import org.jetbrains.annotations.NotNull;

import java.util.UUID;

public class OrderBookCollectCommand extends AbstractIdOrderBookCommand {
    public OrderBookCollectCommand(OrderManager om) {
        super(om);
    }

    @Override
    public boolean onPlayerCommand(@NotNull CommandSender sender, @NotNull Player player, @NotNull Command command, @NotNull String label, @NotNull String[] args) {

        UUID id = parseArgs(args, sender);
        if(id == null) { return true; }

        PlayerInventory inventory = player.getInventory();

        Order order = orderManager.getOrder(id);

        if(order == null) {
            sender.sendMessage("Unable to find order ", id.toString());
            return true;
        }

        if(order.getStage().isPosted()) {
            sender.sendMessage("Unable to collect order. Order is currently posted and has not made it's trade");
            return true;
        }

        if(order.getStage().isCompleted()) {
            sender.sendMessage("Unable to collect order. Order is already completed");
            return true;
        }

        if(!TradeInventoryManager.hasBookSet(order, inventory)){
            sender.sendMessage("You do not have the required trade books");
            return true;
        }

        TradeInventoryManager.takeBookSet(order, inventory);

        TradeInventoryManager.giveBuyItems(order, inventory);

        if( !order.setCompleted() ){
            assert OrderBook.getInstance() != null;
            OrderBook.getInstance().getLogger().warning("Somehow failed to set order to completed even after checking the stage");
        }

        return true;
    }
}

File: src/main/java/orderbook/orderbook/commands/OrderBookCollectCommand.kt
package orderbook.orderbook.commands

import orderbook.orderbook.*
import orderbook.orderbook.log.logComplete
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class OrderBookCollectCommand(plugin: OrderBook) : AbstractIdOrderBookCommand(plugin) {
    override fun onPlayerCommand(
        sender: CommandSender,
        player: Player,
        ignoredCommand: Command,
        ignoredLabel: String,
        args: Array<String>
    ): Boolean {
        val id = parseArgs(args, sender) ?: return true
        val inventory = player.inventory
        val order = plugin.orderManager?.getOrder(id)
        order ?: run {
            sender.sendMessage("Unable to find order ", id.toString())
            return true
        }

        if (!order.stage.isMatched) {
            sender.sendMessage("Unable to collect order. Order is not matched")
            return true
        }

        if (!(inventory hasBookSet order)) {
            sender.sendMessage("You do not have the required trade books")
            return true
        }
        inventory takeBookSet order
        inventory giveBuyItems order
        order.setCompleted()

        plugin.logger.logComplete(player.name, order.sellItem, order.buyItem, order.id)
        return true
    }
}